### PR TITLE
Record Video Functionality

### PR DESCRIPTION
**Pull request by vipulsharma144**
Record Video
Save Video To cache 
Stop Video Record and returns recorded video path

### CODE REVIEW

**Comment by westonganger** [comment]
@vipulsharma144 this is looking pretty good. Can you resolve the conflicts?

**Comment by vipulsharma144** [comment]
@westonganger Conflicts are resolved.

**Comment by westonganger** [comment]
This is a pretty large change. Do you have any additional comments before I start review?

**Comment by vipulsharma144** [comment]
@westonganger  not much to comment on this .Please help in the review .

File: src/ios/CameraPreview.m [hunk 1]
```diff
@@ -455,26 +467,164 @@ - (void) takePicture:(CDVInvokedUrlCommand*)command {
...
-        pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"Camera not started"];
-        [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
+        pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"Session not started"];
+    }
+    
+    [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
+}
+
+-(NSString*)getFileName
+{
```
Comment (westonganger) [inline src/ios/CameraPreview.m]: Move starting bracket to method line ie. `methodName {`

File: src/ios/CameraPreview.m [hunk 2]
```diff
@@ -455,26 +467,164 @@ - (void) takePicture:(CDVInvokedUrlCommand*)command {
...
+    while ([fileManager fileExistsAtPath:tempPath]) {
+        tempPath = [NSString stringWithFormat:@"%@%@_%i%@", libPath, token, fileNameIncrementer, VideoFileExtension];
+        fileNameIncrementer++;
+    }
+
+    return tempPath;
+}
+
+-(NSString*)getLibraryPath
+{
```
Comment (westonganger) [inline src/ios/CameraPreview.m]: Move starting bracket to method line ie. `methodName {`

File: src/ios/CameraPreview.m [hunk 3]
```diff
@@ -455,26 +467,164 @@ - (void) takePicture:(CDVInvokedUrlCommand*)command {
...
+    NSArray *lib = NSSearchPathForDirectoriesInDomains(NSLibraryDirectory, NSUserDomainMask, YES);
+    NSString *library = [lib objectAtIndex:0];
+    return [NSString stringWithFormat:@"%@/NoCloud/", library];
+}
+
+-(NSString*)getCachePath
+{
```
Comment (westonganger) [inline src/ios/CameraPreview.m]: Move starting bracket to method line ie. `methodName {`

File: src/ios/CameraPreview.m [hunk 4]
```diff
@@ -455,26 +467,164 @@ - (void) takePicture:(CDVInvokedUrlCommand*)command {
...
+    NSString* cachePath = NSSearchPathForDirectoriesInDomains(NSCachesDirectory, NSUserDomainMask, YES)[0];
+    return [NSString stringWithFormat:@"%@/", cachePath];
+}
+
+-(AVCaptureDevice *)getCamera: (NSString *)camera
+{
```
Comment (westonganger) [inline src/ios/CameraPreview.m]: Move starting bracket to method line ie. `methodName {`

File: src/ios/CameraPreview.m [hunk 5]
```diff
@@ -455,26 +467,164 @@ - (void) takePicture:(CDVInvokedUrlCommand*)command {
...
+        else if ([camera caseInsensitiveCompare:@"BACK"] == NSOrderedSame)
+        {
+            if (device.position == AVCaptureDevicePositionBack )
+            {
+                captureDevice = device;
+                break;
+            }
+        }
+        else
+        {
```
Comment (westonganger) [inline src/ios/CameraPreview.m]: Fix all `{` bracket positions in this method.

File: typescript/CameraPreview.d.ts [hunk 6]
```diff
@@ -39,40 +39,45 @@ declare module 'cordova-plugin-camera-preview' {
...
+ getExposureCompensation(onSuccess?: CameraPreviewSuccessHandler, onError?: CameraPreviewErrorHandler): void;
+ setExposureCompensation(exposureCompensation?: number, onSuccess?: CameraPreviewSuccessHandler, onError?: CameraPreviewErrorHandler): void;
+ getExposureCompensationRange(onSuccess?: CameraPreviewSuccessHandler, onError?: CameraPreviewErrorHandler): void;
+ getSupportedWhiteBalanceModes(onSuccess?: CameraPreviewSuccessHandler, onError?: CameraPreviewErrorHandler): void;
+ getSupportedWhiteBalanceMode(onSuccess?: CameraPreviewSuccessHandler, onError?: CameraPreviewErrorHandler): void;
+ setWhiteBalanceMode(whiteBalanceMode?: CameraPreviewWhiteBalanceMode|string, onSuccess?: CameraPreviewSuccessHandler, onError?: CameraPreviewErrorHandler): void;
+ onBackButton(onSuccess?: CameraPreviewSuccessHandler, onError?: CameraPreviewErrorHandler): void;
+ getBlob(path: string, onSuccess?: CameraPreviewSuccessHandler, onError?: CameraPreviewErrorHandler): void;
+ getCameraCharacteristics(onSuccess?: CameraPreviewSuccessHandler, onError?: CameraPreviewErrorHandler): void;
+
```
Comment (westonganger) [inline typescript/CameraPreview.d.ts]: Remove empty line.

File: typescript/CameraPreview.d.ts [hunk 7]
```diff
@@ -39,40 +39,45 @@ declare module 'cordova-plugin-camera-preview' {
...
   }
+
```
Comment (westonganger) [inline typescript/CameraPreview.d.ts]: Remove empty line.

File: typescript/CameraPreview.d.ts [hunk 8]
```diff
@@ -39,40 +39,45 @@ declare module 'cordova-plugin-camera-preview' {
...
-   getExposureCompensation(onSuccess?: CameraPreviewSuccessHandler, onError?: CameraPreviewErrorHandler): void;
-   setExposureCompensation(exposureCompensation?: number, onSuccess?: CameraPreviewSuccessHandler, onError?: CameraPreviewErrorHandler): void;
-   getExposureCompensationRange(onSuccess?: CameraPreviewSuccessHandler, onError?: CameraPreviewErrorHandler): void;
-   getSupportedWhiteBalanceModes(onSuccess?: CameraPreviewSuccessHandler, onError?: CameraPreviewErrorHandler): void;
-   getSupportedWhiteBalanceMode(onSuccess?: CameraPreviewSuccessHandler, onError?: CameraPreviewErrorHandler): void;
-   setWhiteBalanceMode(whiteBalanceMode?: CameraPreviewWhiteBalanceMode|string, onSuccess?: CameraPreviewSuccessHandler, onError?: CameraPreviewErrorHandler): void;
-   onBackButton(onSuccess?: CameraPreviewSuccessHandler, onError?: CameraPreviewErrorHandler): void;
-   getBlob(path: string, onSuccess?: CameraPreviewSuccessHandler, onError?: CameraPreviewErrorHandler): void;
-   getCameraCharacteristics(onSuccess?: CameraPreviewSuccessHandler, onError?: CameraPreviewErrorHandler): void;
+
```
Comment (westonganger) [inline typescript/CameraPreview.d.ts]: Remove empty line.

File: src/ios/CameraSessionManager.m [hunk 9]
```diff
@@ -418,7 +456,7 @@ - (NSArray *)getExposureModes {
     [exposureModes addObject:@"auto"];
   };
   if ([videoDevice isExposureModeSupported:2]) {
-    [exposureModes addObject:@"continuous"];
+    [exposureModes addObject:@"cotinuous"];
```
Comment (westonganger) [inline src/ios/CameraSessionManager.m]: Please fix this. It should say `continuous`

File: src/ios/CameraSessionManager.m [hunk 10]
```diff
@@ -153,6 +153,44 @@ - (void) setupSession:(NSString *)defaultCamera completion:(void(^)(BOOL started
...
+}
+
+-(void)stopRecordVideo {
+    self.isRecording = false;
+//    [self.assetWriterMyData finishWriting];
+    [self.assetWriterMyData finishWritingWithCompletionHandler:^(){
+        NSLog (@"finished writing");
+    }];
+    //[self.session removeOutput:self.videoFileOutput];
+    //self.videoFileOutput = nil;
```
Comment (westonganger) [inline src/ios/CameraSessionManager.m]: Remove comments and fix indentation.

File: src/ios/CameraSessionManager.m [hunk 11]
```diff
@@ -153,6 +153,44 @@ - (void) setupSession:(NSString *)defaultCamera completion:(void(^)(BOOL started
   });
 }
 
+-(void) startRecordVideo:(NSURL *) fileUrl {
+
```
Comment (westonganger) [inline src/ios/CameraSessionManager.m]: Remove empty line.

File: src/ios/CameraSessionManager.m [hunk 12]
```diff
@@ -153,6 +153,44 @@ - (void) setupSession:(NSString *)defaultCamera completion:(void(^)(BOOL started
...
+    NSDictionary *outputSettings = [NSDictionary dictionaryWithObjectsAndKeys:
+                                    [NSNumber numberWithInt:720], AVVideoWidthKey, [NSNumber numberWithInt:1280], AVVideoHeightKey, AVVideoCodecH264, AVVideoCodecKey, nil];
+    self.assetWriterInput = [AVAssetWriterInput  assetWriterInputWithMediaType:AVMediaTypeVideo outputSettings:outputSettings];
+    
+    self.pixelBufferAdaptor = [[AVAssetWriterInputPixelBufferAdaptor alloc]initWithAssetWriterInput:self.assetWriterInput sourcePixelBufferAttributes:[NSDictionary dictionaryWithObjectsAndKeys:
+                [
+                    NSNumber numberWithInt:kCVPixelFormatType_32BGRA],kCVPixelBufferPixelFormatTypeKey,nil]
+               ];
+    
+
```
Comment (westonganger) [inline src/ios/CameraSessionManager.m]: Remove empty line.

File: src/ios/CameraSessionManager.m [hunk 13]
```diff
@@ -153,6 +153,44 @@ - (void) setupSession:(NSString *)defaultCamera completion:(void(^)(BOOL started
...
+                                    [NSNumber numberWithInt:720], AVVideoWidthKey, [NSNumber numberWithInt:1280], AVVideoHeightKey, AVVideoCodecH264, AVVideoCodecKey, nil];
+    self.assetWriterInput = [AVAssetWriterInput  assetWriterInputWithMediaType:AVMediaTypeVideo outputSettings:outputSettings];
+    
+    self.pixelBufferAdaptor = [[AVAssetWriterInputPixelBufferAdaptor alloc]initWithAssetWriterInput:self.assetWriterInput sourcePixelBufferAttributes:[NSDictionary dictionaryWithObjectsAndKeys:
+                [
+                    NSNumber numberWithInt:kCVPixelFormatType_32BGRA],kCVPixelBufferPixelFormatTypeKey,nil]
+               ];
+    
+    
+    /* Asset writer with MPEG4 format*/
```
Comment (westonganger) [inline src/ios/CameraSessionManager.m]: Remvoe comment.

File: src/ios/CameraSessionManager.m [hunk 14]
```diff
@@ -137,7 +137,7 @@ - (void) setupSession:(NSString *)defaultCamera completion:(void(^)(BOOL started
 
       AVCaptureVideoDataOutput *dataOutput = [[AVCaptureVideoDataOutput alloc] init];
       if ([self.session canAddOutput:dataOutput]) {
-        self.dataOutput = dataOutput;
+          self.dataOutput = dataOutput;
```
Comment (westonganger) [inline src/ios/CameraSessionManager.m]: Fix indentation.

File: src/ios/CameraRenderController.m [hunk 15]
```diff
@@ -215,6 +231,15 @@ -(void)captureOutput:(AVCaptureOutput *)captureOutput didOutputSampleBuffer:(CMS
     [self.ciContext drawImage:croppedImage inRect:dest fromRect:[croppedImage extent]];
     [self.context presentRenderbuffer:GL_RENDERBUFFER];
     [(GLKView *)(self.view)display];
+
```
Comment (westonganger) [inline src/ios/CameraRenderController.m]: Remove empty line.

File: src/ios/CameraRenderController.m [hunk 16]
```diff
@@ -157,7 +157,23 @@ - (void) applicationEnteredForeground:(NSNotification *)notification {
...
+//    if(self.sessionManager.assetWriterInput.readyForMoreMediaData) {
+//        [self.sessionManager.pixelBufferAdaptor appendPixelBuffer:imageBuffer withPresentationTime:CMTimeMake(frameNumber, 25)];
+//    }
+//
+//    frameNumber++;
+//}
+
 -(void)captureOutput:(AVCaptureOutput *)captureOutput didOutputSampleBuffer:(CMSampleBufferRef)sampleBuffer fromConnection:(AVCaptureConnection *)connection {
+    
+
```
Comment (westonganger) [inline src/ios/CameraRenderController.m]: Remove empty lines.

File: src/ios/CameraRenderController.m [hunk 17]
```diff
@@ -157,7 +157,23 @@ - (void) applicationEnteredForeground:(NSNotification *)notification {
       });
 }
 
+//-(void) captureOutput:(AVCaptureOutput *)captureOutput didOutputSampleBuffer:(CMSampleBufferRef)sampleBuffer fromConnection:(AVCaptureConnection *)connection{
```
Comment (westonganger) [inline src/ios/CameraRenderController.m]: Remove unused/comment method.

File: src/ios/CameraPreview.m [hunk 18]
```diff
@@ -738,52 +889,16 @@ - (void) invokeTakePicture:(CGFloat) width withHeight:(CGFloat) height withQuali
...
-
-    // generate unique file name
-    int i = 1;
-    do {
-        filePath = [NSString stringWithFormat:@"%@/%@%04d.%@", tmpPath, TMP_IMAGE_PREFIX, i++, extension];
-    } while ([fileMgr fileExistsAtPath:filePath]);
-    
-    return filePath;
-}
-
```
Comment (westonganger) [inline src/ios/CameraPreview.m]: Please re-add getTempDirectoryPath and getTempFilePath. I believe these are used for takeSnapshot.

File: plugin.xml [hunk 19]
```diff
@@ -23,8 +25,7 @@
     <source-file src="src/android/camera_theme.xml" target-dir="res/values" />
     <source-file src="src/android/camera_ids.xml" target-dir="res/values" />
 
-    <preference name="ANDROID_SUPPORT_LIBRARY_VERSION" default="26+"/>
-    <framework src="com.android.support:exifinterface:$ANDROID_SUPPORT_LIBRARY_VERSION" />
+    <framework src="com.android.support:exifinterface:25+" />
```
Comment (westonganger) [inline plugin.xml]: Please remove this change for exifInterface version and revert to previous implementation.

File: README.md [hunk 20]
```diff
@@ -410,32 +422,65 @@ CameraPreview.setExposureCompensation(3);
...
 ```javascript
-CameraPreview.getCameraCharacteristics(function(characteristics){
+CameraPreview.getCameraCharacteristics(function(characteristics) {
   console.log(characteristics);
 });
 ```
 
+### backgroundVideoStart(cb, [errorCallback])
+
+_Currently this feature is for Android only._
```
Comment (westonganger) [inline README.md]: Is this actually android only? I am pretty sure the iOS is implemented too.

File: README.md [hunk 21]
```diff
@@ -410,32 +422,65 @@ CameraPreview.setExposureCompensation(3);
...
+
+```javascript
+CameraPreview.backgroundVideoStart(function(result) {
+  console.log(result);
+});
+```
+
+### backgroundVideoStop(cb, [errorCallback])
+
+_Currently this feature is for Android only._
```
Comment (westonganger) [inline README.md]: Is this actually android only? I am pretty sure the iOS is implemented too.

File: README.md [hunk 22]
```diff
@@ -410,32 +422,65 @@ CameraPreview.setExposureCompensation(3);
...
 <info>Get the characteristics of all available cameras. Returns a JSON object representing the characteristics of all available cameras.</info><br/>
 
 ```javascript
-CameraPreview.getCameraCharacteristics(function(characteristics){
+CameraPreview.getCameraCharacteristics(function(characteristics) {
   console.log(characteristics);
 });
 ```
 
+### backgroundVideoStart(cb, [errorCallback])
```
Comment (westonganger) [inline README.md]: Lets rename this method to `recordVideoStart`

File: README.md [hunk 23]
```diff
@@ -410,32 +422,65 @@ CameraPreview.setExposureCompensation(3);
...
+
+<info>Record Video and start saving to the cache</info><br/>
+
+```javascript
+CameraPreview.backgroundVideoStart(function(result) {
+  console.log(result);
+});
+```
+
+### backgroundVideoStop(cb, [errorCallback])
```
Comment (westonganger) [inline README.md]: Lets rename this method to recordVideoStop

**Comment by westonganger** [comment]
@vipulsharma144 please follow up with my comments so we can keep the ball rolling on this PR.

**Comment by vipulsharma144** [comment]
@westonganger whoa 🤩 !! Thanks for the detailed review I am on it . Will update asap.

**Comment by westonganger** [comment]
Alright this looks good enough now. Thanks @vipulsharma144 for your hard work on this PR.

File: README.md (hunk 21 above)
Comment (mgh0010) [reply README.md:468]: Was iOS ever confirmed?

Comment (westonganger) [reply README.md:468]: No it wasnt actually implemented. PR wanted for iOS support.

**Pull request merged**